### PR TITLE
Implement MSC2965 action parameter

### PR DESCRIPTION
**Pull request by kerryarchibald**
Fixes https://github.com/matrix-org/matrix-authentication-service/issues/1550

Redirects to session detail when query param `action=session_end` is encountered.

https://github.com/matrix-org/matrix-authentication-service/assets/3055605/e1ec182e-7310-4127-8434-bad217811b19

### CODE REVIEW

File: frontend/src/Router.tsx [hunk 1]
```diff
@@ -169,14 +143,32 @@ export const routeAtom = atom(
...
+ * Get route from path
+ * @returns Route
+ */
+const useRouteWithRedirect = (): Route => {
+  const location = useAtomValue(locationAtom);
+  const redirect = getRouteActionRedirection(location);
+
+  const [route, setRoute] = useAtom(routeAtom);
+  if (redirect) {
+    setRoute(redirect.route, redirect.searchParams);
```
Comment (sandhose) [inline frontend/src/Router.tsx:166]: I think this should be in a useEffect, because you're not supposed to do state changes in the render?

Comment (Johennes) [reply frontend/src/Router.tsx:166]: Kerry won't see this before Monday, sadly, and we have to land this PR today. @t3chguy could you help us out by giving advice?

Comment (sandhose) [reply frontend/src/Router.tsx:166]: I can take care of finishing that up, alongside the server side of this

Comment (t3chguy) [reply frontend/src/Router.tsx:166]: Yes I believe it should be in a useEffect

**Comment by sandhose** [comment]
Alright, I: 
 - refactored a bit the router
 - did the backend part of the redirect so that params are kept through the login
 - implemented the other `action` defined in the MSC on the frontend

**Pull request merged**